### PR TITLE
PAI review: install fixes, numerical correctness, MLA cache, generate hardening (+ roadmap)

### PR DESCRIPTION
**Pull request by baseflux**
External code review conducted via the PAI Algorithm (v3.8.0, DETERMINED effort). Eight specialist reviewers ran in parallel — correctness, performance, adversarial stress, maintainability, testing, python-style, architecture, and PyTorch 2026 best practices. 30 findings total; this PR lands Tier 0 (install-blockers) and Tier 1 (correctness critical).

**All 81 tests pass.** Full review + prioritized follow-up roadmap is in [`PAI_REVIEW_2026-04.md`](https://github.com/baseflux/OpenMythos/blob/pai-review-2026-04/PAI_REVIEW_2026-04.md).

## Nine commits, independently splittable

1. **`fix(install)`** — `torch==2.11.0` pin was a nonexistent version; relaxed to `>=2.4,<3`. `__all__` listed `load_tokenizer`/`get_vocab_size` which were never defined (\`from open_mythos import *\` → AttributeError). README quickstart called undefined `mythos_7b()`; fixed to `mythos_3b()`. Deferred `transformers` import so \`import open_mythos\` no longer ImportErrors without transformers. Added loguru (used by training script) + pytest config + 3.11/3.12 classifiers.

2. **`feat(config): __post_init__ validation`** — previously invalid combos (typo'd \`attn_type\`, non-divisible heads, odd head_dim, topk>n_experts, max_loop_iters<1) silently fell through or crashed deep in forward. Now raises \`ValueError\` at construction. \`attn_type\` typed \`Literal[\"gqa\", \"mla\"]\`.

3. **`fix(numerics)`** — four silent bugs:
   - \`LTIInjection.get_A\` bf16 underflow: \`exp(-10)\` → 0, then \`exp(-0)\` = 1.0, silently making ρ(A)=1 (marginal stability, not strictly <1). Now computes in fp32 with tightened \`clamp(-10, 10)\`.
   - \`loop_index_embedding\` bf16 frequency collapse (7-bit mantissa can't distinguish adjacent k indices). Now fp32 compute, cast once at the end.
   - \`OpenMythos.__init__\` ran \`_init_weights\` *after* \`head.weight = embed.weight\`, initializing the shared tensor twice. Moved tying after init.
   - \`RecurrentBlock.forward\`: positions that never halted had \`cumulative_p < threshold\`, so their h_out was scaled by <1 while halted positions summed to exactly 1. After the loop, flush remaining mass onto still-running positions.

4. **`feat(generate)`** — \`forward\` now raises on T=0 and \`start_pos+T > max_seq_len\` (previously silently produced garbage). \`generate()\` calls \`self.eval()\` for the duration (restoring prior mode on exit), accepts \`eos_token_id\` with per-row stopping, clamps \`top_k\` to vocab_size.

5. **`fix(mla): cache shared k_rope once`** — \`MLAttention\` was expanding \`k_rope\` to \`(B, T, n_heads, rope_dim)\` via \`.expand().contiguous()\` before caching, storing \`n_heads\` identical copies per token. At \`n_heads=16\` the KV cache was 16× larger than DeepSeek-V2 MLA specifies, negating the memory savings that motivate MLA. Now caches the shared \`(B, T, rope_dim)\` once; per-head expansion happens at compute time via a broadcast view (cost-free, not a copy).

6. **`test: PAI regression suite + Cyrillic fix`** — 14 regression tests covering every fix above. Also fixes \`TestOpenMythosMLА\` (Cyrillic А U+0410) → \`TestOpenMythosMLA\` (ASCII); previously grep missed it.

7. **`fix(tests): slice freqs_cis to T`** — 13 unit tests in TestGQAttention/MLAttention/TransformerBlock/RecurrentBlock passed the full \`(max_seq_len, head_dim//2)\` freqs_cis table into attention with \`T < max_seq_len\`, which broadcast-fails in \`apply_rope\`. The tests were silently broken because the pre-deferred \`transformers\` import meant \`test_main.py\` couldn't load at all without transformers installed. Fixed by slicing \`[:T]\` at every call site.

8. **`chore: gitignore uv.lock`** — tool artifact from test run.

9. **`docs: PAI review roadmap`** — 502-line prioritized follow-up document. Tier 2 (9 performance items: SDPA swap, MoE grouped-GEMM, KV cache preallocation, gradient checkpointing, router-bias update loop, RoPE real-pair, \`torch.compile\`, tokenization batching), Tier 3 (9 correctness-medium), Tier 4 (10-point architecture debt including the orphan \`moda.py\`), Tier 5 (16 concrete tests to add). Each item keyed to reviewer finding with fix sketch + expected impact.

## Verification

\`\`\`
pytest test_main.py tests/test_pai_regressions.py tests/test_rope_debug.py
...  (81 tests)  ...
81 passed in 1.14s
\`\`\`

## Reviewer artifacts

All eight reviewers returned structured findings at DETERMINED effort. Severity distribution across all 30 findings: 4 CRITICAL, 13 HIGH, 9 MEDIUM, 4 LOW. Landed 3 CRIT, 4 HIGH, 4 MED, 1 LOW; remainder documented as tiered follow-up roadmap.

Happy to split this into focused PRs if one bundle is too heavy — each commit is independent and cleanly scoped.

### CODE REVIEW

**Comment by baseflux** [comment]
Correction on commit 1's rationale: I claimed torch 2.11.0 was "a nonexistent version on PyPI." It exists (installed cleanly on my end during smoke testing). The reviewer-agent that surfaced that finding had stale training data. The change is still the right call for a different reason — exact pins in library deps (\`torch = "2.11.0"\`) break downstream users whenever PyTorch releases a new minor; a range pin (\`>=2.4,<3\`) serves them better. Apologies for the misleading justification in the original PR body and in PAI_REVIEW_2026-04.md.

Other 8 commits unaffected.

**Comment by Qodo-Free-For-OSS** [comment]
Hi, MythosConfig.__post_init__ enforces n_heads % n_kv_heads == 0 even when attn_type="mla", but MLA does not use n_kv_heads. This rejects otherwise valid MLA configurations at construction time for no functional reason.

**Severity:** action required | **Category:** correctness

**How to fix:** Gate check by attn_type

**Agent prompt to fix** - you can give this to your LLM of choice:

> ## Issue description
> `MythosConfig.__post_init__` currently enforces `n_heads % n_kv_heads == 0` for *all* configs, but this constraint is only required when `attn_type == "gqa"`. For `attn_type == "mla"`, `n_kv_heads` is ignored by the attention implementation, so the validation incorrectly rejects valid MLA configs.
> 
> ## Issue Context
> - `TransformerBlock` selects `MLAttention` when `cfg.attn_type == "mla"`.
> - `MLAttention` does not read `cfg.n_kv_heads`.
> - `GQAttention` does rely on `n_kv_heads` via `groups = n_heads // n_kv_heads`.
> 
> ## Fix Focus Areas
> - open_mythos/main.py[76-114]
> 
> ### Suggested change
> Wrap the `n_heads % n_kv_heads` check with `if self.attn_type == "gqa":` (or equivalently skip it when `attn_type == "mla"`). Optionally, adjust the error message to remove “for GQA” if you keep the check unconditional (not recommended).

---
*Found by [Qodo](https://qodo.ai) code review*

**Comment by baseflux** [comment]
**Update**: pushed a 10th commit landing two Tier 2 perf items from the roadmap:

- \`F.scaled_dot_product_attention\` in both GQAttention and MLAttention (drops manual softmax(QK^T)V; auto-dispatches FlashAttention-2/3 on CUDA bf16/fp16; no longer materializes the full (B,H,T,S) attention matrix — ~1GB/layer saved at production config). Passes \`is_causal=True\` when mask=None and kv_cache is empty so the fused-causal kernel fires in the common case.
- \`MythosConfig.gradient_checkpointing\` flag (default False). When True and training w/o a KV cache, wraps the inner TransformerBlock with \`torch.utils.checkpoint(..., use_reentrant=False)\`. Saves ~sqrt(n_loops) activation memory. No-op in eval mode.

Added 2 new regression tests (SDPA forward correctness, checkpointing numerical equivalence). All 16 PAI regressions + all 81 existing tests green.

Still the same branch; same reviewer roadmap still applies for Tier 2.3+.
